proposals/0000-unify-type-level-literals.rst

```diff
@@ -0,0 +1,156 @@
+
+::
+
+    safePackBS :: forall str. (ValidBS str) => BS.ByteString
+    safePackBS = BS.pack $ symbolVal (Proxy :: Proxy str)
+
+    validBS, invalidBS :: BS.ByteString
+    validBS = safePackBS @"foo"
+    -- works!
+    invalidBS = safePackBS @"中国語"
+    -- fails with error: '\20013' is not a single-byte character.
+    -- Gives clean, helpful error message for invalid input at compile time.
+
+Turning compile-time problems into compile-time errors is one of the main reasons that I use Haskell, and this value-transparency would allow that to be done cleanly in more cases, and the opacity of ``Symbol``\s is not the only restriction that is unfortunately limiting.
+
+This is largely motivated by the discussion on ghc-proposals/ghc-proposals#124 into a way to use type-level verification to make it safe to implement overloaded literals for the many types for which it is currently not possible, but I feel that this is both a substantial enough change and a potentially useful enough feature to be its own proposal rather than something added to a continuation of that work.
+
+Proposed Change Specification
+-----------------------------
+The following is relative to the current behavior of ``DataKinds``.
+
+Grammar changes
+^^^^^^^^^^^^^^^
+The lexical syntax would gain the following new case:
+
+::
+
+    signedint -> ['+'|'-'] integer
+
+The ``atype`` non-terminal would gain new cases as follows:
+
+::
+
+    atype -> ...
+           | string
+           | char
+           | integer
+           | signedint
+           | float
+
+This is relative to the specification, not the current effects of ``DataKinds``.
+
+Semantics
+^^^^^^^^^
+The kinds of type-level literals will be as follows:
+
+::
+
+    "foo" :: String
+    'f' :: Char
+    -123 :: Integer
+    +123 :: Integer
+    123 :: Natural
+    3.14 :: Rational
+
+While the ``Rational`` literals are the most dubious, they require very minimal changes (since ``Ratio Integer`` will now work correctly "for free") and are needed to satisfy the original impetus for this change.
+
+The case of positive ``Integer`` literals is challenging. The syntax above is not great, but the obvious alternatives are a subkinding relation or an explicit conversion and I'd consider both of those to be worse.
+
+To handle the legacy case, ``Symbol`` and ``Nat`` will become aliases for ``String`` and ``Natural``, respectively.
+
+All of the mathematical type families in ``GHC.TypeNats`` will be generalized to be poly-kinded so that they work for these new numeric kinds, with the exception of a ``Rational`` implementation for  ``Div``, ``Mod``, and ``Log2``. ``/`` and ``%`` will be added for ``Rational``. Put more explicitly, their interfaces will be as listed in the block below, where ``a`` means any of ``Integer``, ``Natural``, or ``Rational`` and ``b`` means either ``Integer`` or ``Rational``.
+
+::
+
+    type family Add a (m :: a) (n :: a) :: a
```

Comment:
You can just remove this `a` parameter and everything will still work (it will become an invisible parameter). Then there will be no need for all the wrappers. See also my comment on `Cmp`, it's the same thing.